### PR TITLE
Refactor the Flask application to use an app factory

### PR DESCRIPTION
**Pull request by jeremycline**
This refactor removes the module-level ``anitya.app.APP`` object and replaces it with a function that will create and configure a complete Flask application. This, among other things, makes testing easier since each test can create, configure, and modify its own Flask application object. Check out http://flask.pocoo.org/docs/0.12/patterns/appfactories/ for an overview of the general pattern.

I did my best to break this up into a series of commits that are atomic and make sense. They're probably not as neat as I'd like, but hopefully they make this PR a little less painful to review (sorry!).

### CODE REVIEW

File: anitya/app.py [hunk 1]
```diff
@@ -79,37 +77,12 @@ def create(config=None):
             mail_admin=app.config.get('ADMIN_EMAIL')
         ))
 
-
```
Comment (bowlofeggs) [inline anitya/app.py]: These space changes are good, but would be better to put into an atomic commit.

File: anitya/app.py [hunk 2]
```diff
@@ -50,26 +49,20 @@ def create(config=None):
 
     if config is None:
         config = anitya_config
-
```
Comment (bowlofeggs) [inline anitya/app.py]: This change would be better to do in its own commit as well.

File: anitya/app.py [hunk 3]
```diff
@@ -50,26 +49,20 @@ def create(config=None):
...
 
-    # Mark login handlers
-    app.oid.loginhandler(ui.login)
-    app.oid.loginhandler(ui.fedora_login)
-    app.oid.loginhandler(ui.yahoo_login)
-    app.oid.loginhandler(ui.google_login)
-
     app.before_request(check_auth)
     app.teardown_request(shutdown_session)
+
```
Comment (bowlofeggs) [inline anitya/app.py]: Here as well.

File: anitya/ui.py [hunk 4]
```diff
@@ -10,8 +11,26 @@
...
 from anitya.lib import utilities
+from anitya.lib.model import Session as SESSION
+
+
+ui_blueprint = flask.Blueprint(
+    'anitya_ui', __name__, static_folder='static', template_folder='templates')
+
+
+def login_required(function):
+    ''' Flask decorator to retrict access to logged-in users. '''
```
Comment (jeremycline) [inline anitya/ui.py]: Normally, I'd be all for this (you know how much I like documentation), but I also plan on replacing this with [flask-login's version](https://flask-login.readthedocs.io/en/latest/#flask_login.login_required) because it integrates nicely with [python-social-auth](http://python-social-auth.readthedocs.io/en/latest/configuration/flask.html)

**Comment by ncoghlan** [comment]
I don't have an opinion on the exact details of the changes, but a big +1 for the general idea (especially if it gets you closer to enabling full social auth support).
